shared/subspace-data-retrieval/src/segment_fetcher.rs

```diff
@@ -85,66 +80,87 @@ where
+                .split_at_checked(batch_size)
+                .unwrap_or((piece_indexes, &[]));
+            // Invariant: the number of remaining piece indexes gets smaller.
+            piece_indexes = remaining_piece_indexes;
+
+            let stream = piece_getter.get_pieces(batch_indexes.iter().cloned()).await;
+            match stream {
+                Ok(stream) => {
+                    piece_streams.push(stream);
+                    pieces_pending += batch_size;
+                }
+                Err(error) => {
+                    // A single batch failure isn't fatal.
+                    debug!(
+                        ?error,
+                        %segment_index,
+                        batch_size,
+                        pieces_pending,
+                        pieces_received,
+                        pieces_needed = RecordedHistorySegment::NUM_RAW_RECORDS,
+                        "Segment piece getter batch failed"
+                    );
+                }
+            }
+        }
 
-    while let Some(maybe_piece) = received_segment_pieces.next().await {
-        let Some((piece_index, piece)) = maybe_piece else {
-            continue;
+        // Poll all the batches concurrently, getting all finished pieces.
+        let piece_responses = stream::iter(&mut piece_streams)
+            .flatten_unordered(None)
+            .ready_chunks(RecordedHistorySegment::NUM_RAW_RECORDS)
+            .next()
+            .await;
+
+        let Some(piece_responses) = piece_responses else {
+            // All streams have finished, perhaps abnormally, so reset the number of pending pieces.
+            // Invariant: the number of pending pieces gets smaller.
+            pieces_pending = 0;
+            continue 'fetcher;
         };
 
-        segment_pieces
-            .get_mut(piece_index.position() as usize)
-            .expect("Piece position is by definition within segment; qed")
-            .replace(piece);
+        // Process the piece responses.
+        'processor: for maybe_piece in piece_responses {
+            // Invariant: the number of pending pieces gets smaller.
+            pieces_pending -= 1;
+
+            let (piece_index, Ok(Some(piece))) = maybe_piece else {
+                continue 'processor;
+            };
 
-        pieces_received += 1;
+            segment_pieces
+                .get_mut(piece_index.position() as usize)
+                .expect("Piece position is by definition within segment; qed")
+                .replace(piece);
 
-        if pieces_received >= RecordedHistorySegment::NUM_RAW_RECORDS {
-            trace!(%segment_index, "Received half of the segment.");
-            break;
+            pieces_received += 1;
+
+            if pieces_received >= RecordedHistorySegment::NUM_RAW_RECORDS {
+                trace!(%segment_index, "Received half of the segment.");
+                break 'fetcher;
+            }
```

Comment:
Do we consider a case with insufficient pieces present?

Reply:
When there aren't enough pieces, we go to the top of the `'fetcher:` loop again, and try to get more pieces from the segment.

If we've tried all the pieces in the segment, and can't get enough to reconstruct the segment, we end the `'fetcher:` loop, and return an error.

That error isn't visible in this diff, it's on line 159/175, about 10 lines below this line.